toc/rfc/rfc-draft-bindable-secrets.md

```diff
@@ -0,0 +1,254 @@
+    "self": {
+      "href": "https://api.example.org/v3/service_instances/c89b3280-fe8d-4aa0-a42e-44465bb1c61c"
+    },
+    "space": {
+      "href": "https://api.example.org/v3/spaces/5a84d315-9513-4d74-95e5-f6a5501eeef7"
+    },
+    "secret_bindings": {
+      "href": "https://api.example.org/v3/secret_bindings?secret_guids=4f00b75b-3455-48f1-ba4a-f3b82ebad943"
+    },
+    "value": {
+      "href": "https://api.example.org/v3/secrets/4f00b75b-3455-48f1-ba4a-f3b82ebad943/value"
+    }
+  }
+}
+```
+
+The `type` field corresponds with a [CredHub credential type](https://docs.cloudfoundry.org/credhub/credential-types.html). To start, `secret`s will support the `value` and `json` credential types. In the future, this can be expanded to other CredHub-supported types like `certificate`, to support additional use cases.
+
+#### Example `secret_binding` Object
+
+```json
+{
+  "guid": "dde5ad2a-d8f4-44dc-a56f-0452d744f1c3",
+  "created_at": "2015-11-13T17:02:56Z",
+  "updated_at": "2016-06-08T16:41:26Z",
+  "name": "secret-binding-name",
+  "mount_path": "/etc/my-conf/",
+  "metadata": {
+    "annotations": {},
+    "labels": {}
+  },
+  "relationships": {
+    "app": {
+      "data": {
+        "guid": "74f7c078-0934-470f-9883-4fddss5b8f13"
+      }
+    },
+    "secret": {
+      "data": {
+        "guid": "4f00b75b-3455-48f1-ba4a-f3b82ebad943"
+      }
+    }
+  },
+  "links": {
+    "self": {
+      "href": "https://api.example.org/v3/service_credential_bindings/dde5ad2a-d8f4-44dc-a56f-0452d744f1c3"
+    },
+    "secret": {
+      "href": "https://api.example.org/v3/secrets/4f00b75b-3455-48f1-ba4a-f3b82ebad943"
+    },
+    "app": {
+      "href": "https://api.example.org/v3/apps/74f7c078-0934-470f-9883-4fddss5b8f13"
+    }
+  }
+}
+```
+
+The `mount_path` field defines the `tmpfs` directory that will contain files for the `secret`'s content. If a credential has subkeys (such as a [`json` type credential](https://docs.cloudfoundry.org/api/credhub/version/main/#_find_a_credential_by_id_type_json) or [`certificate` type](https://docs.cloudfoundry.org/api/credhub/version/main/#_find_a_credential_by_id_type_certificate)), then files will be made for each key.
+
+For example, consider a `json`-type `secret` with the following `value`:
+
+```json
+"value": {
+  "application.yml" : "---\napplication:\n ...",
+  "secret.txt": "my-secret"
+}
+```
+
+The above `secret` would mount files named `application.yml` and `secret.txt` at the path configured in the binding.
+
+For a `value`-type `secret`, the file name be the `secret_binding`'s name.
+
+### Creating a `secret`
+
+To create a `secret`, users will use a new CLI command: `cf create-secret`.
+
+```console
+cf create-secret SECRET_NAME SECRET_TYPE -p <CREDENTIALS_INLINE_OR_FILE>
+```
+
+If `SECRET_TYPE` is `value` then contents of `-p` will not be parsed. If `SECRET_TYPE` is `json`, `certificate`, or some other CredHub credential type, then the contents of `-p` MUST be valid JSON and may be parsed.
```

Comment:
**question:** I seems to recall some issues where users wanted to avoid any secrets being exposed in CLI commands or outputs in case they are using the CLI in CI/CD environments.  Not supporting a file for `value` would expose secrets in those cases. I suppose that would only affect bash scripts being run with `+x` and the like. Is that something worth worrying about?

Reply:
Oh I didn't intend to mean that it won't support files for `value` secrets with type `-p`, but rereading this I can see how it implies that. I will rework this sentence (especially if we get rid of the `value` type).

Reply:
**suggestion:** The question above aside, I might be simpler to have two flags with two distinct uses:

`-p` reads inline/literal credentials, no matter the `SECRET_TYPE`
`-f` reads credentials from a file, no matter the `SECRET_TYPE`

This would be analogous the `kubectl` `--from-literal` and `--from-file` flags, and wouldn't require users to remember which `SECRET_TYPE` need to be in files and which don't

Reply:
Cool, that makes sense to me. @Gerg suggested something similar here https://github.com/tcdowney/community/pull/1#discussion_r1794288536 and I initially convinced him with the argument that this mirrors the `cf cups` and `cf bind-service` flag usage, but I'm convinced now that discrete flags is simpler to reason about.

Reply:
> Oh I didn't intend to mean that it won't support files for value secrets with type -p, but rereading this I can see how it implies that. I will rework this sentence (especially if we get rid of the value type).

Ah, got it.  One last thought then: if `-p`supports either a file or inline-literal, then mistyping a file name could result in a secret with an incorrect value (e.g. setting your password to `mysecrets.josn`)

Reply:
Or only accept files, and teach people how to use `<()` 😉

Reply:
Oh, I didn't realize we already had that pattern in the CLI. That is a strong argument

Although I have to admit I first was thinking that we used `-f` for files when applying manifests, so we should use the same pattern here 😄

Reply:
Maybe we can avoid this pattern in CLI v9

Reply:
Updated in ff25954